### ISSUE TITLE
Add a server side buffer limit

### ISSUE THREAD

**Issue by jfheins** [dotnet/aspnetcore#43093]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

We have an internal web application in production that uses ASP.net core SignalR. Our app has dozens of users that might open 4 tabs each. However, some of them use mobile devices that have little processing power.
We send about 100 messages per second to every client to keep them up to date. We can see that the (JS) clients have 100% CPU usage and slowly fall behind real time when the message load is high.

Over 99% of the messages are state messages, i.e. as soon as a newer state is known, all older states of the same widget are irrelevent.
On the backend (.net 6 Application with Kestrel in a K8s cluster, connected via websockets) the memory consumption grows linearly when such a slow client id connected:

![image](https://user-images.githubusercontent.com/5293502/182873797-a3e76765-3b70-44c3-8a6e-f02e4adf744d.png)

While we are also working on the UI itself to increase efficiency, my issue is that one slow client can impact the backend in a dramatic way and degrade the experience for all other clients. I suspect the memory increases because the backend buffers many messages for the slow client. While this makes sense for many scenarios, for our high volume message it is useless to deliver a message that sat in the queue for more than 2 seconds.

### Describe the solution you'd like

I would like to limit the server side buffering. Ideally, this would happen on a per-group basis so that we can designate some groups as important with a bigger buffer and other groups as fire-and-forget so that old, undelivered messages get discarded server side.
A limited buffer would make the memory consumption predictable but a TTL might be more relatable.

So a TTL API could look like this:
```
await _hubContext.Clients.Groups("positions")
.SendAsync("positions", "{x: 1, y: 2}", timeToLive: TimeSpan.FromSeconds(2));
```
We tried to use the cancellation token in a similar way but without success.

A per-client buffer limit would be desirable as well and could be configured in the options on startup.

**Comment by davidfowl** [comment]
1. Have you captured a memory dump and analyzed the memory to verify some of your assumptions about message buffering? That would be the first thing I would do to investigate this issue. 
2. How big are these messages?
3. Have you tried reducing the message rate? Maybe you can do something here where you use a bounded channel to store the most recent values to send to the group? It's basically a second layer of buffering on top before you send the data to the client.
4. There's a safeguard in place to make sure that slow clients can't clog out the entire pipeline. There's [a timeout](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connections.httpconnectiondispatcheroptions.transportsendtimeout?view=aspnetcore-6.0) you can use to control this.
5. There are knobs at the transport level (when you call MapHub) that let you control [the various buffer sizes](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connections.httpconnectiondispatcheroptions?view=aspnetcore-6.0), however, these will not cause back pressure to be applied to at the transport level and will not drop data in the transport. You might be able to combine a smaller buffer size with a cancellation token to achieve the behavior you're trying to accomplish.

**Comment by davidfowl** [comment]
@jfheins can you put the code in a GitHub repro?

**Comment by davidfowl** [comment]
> When I serialize them into a string, strings are the biggest users of memory

Why are you doing this?


> In other words, we do not want to degrade the experience for fast clients to the level where even the slowest device can keep up.

What’s the expectation for the client if it can’t keep up?

**Comment by davidfowl** [comment]
Looking at the sample it doesn't use groups, its broadcasting:

> I have added a CancellationToken with 20s timeout so my expectation is that the process uses approximately 3KiB * 200 messages/sec * 20 sec = 12 MiB additional memory to hold the items

There's overhead beyond your serialized messages that you're not accounting for, you can see that in a memory profile (e.g. https://github.com/dotnet/aspnetcore/issues/41343).

How are you measuring memory usage? Working set? What are your memory usage expectations based on? 

> I have to look deeper into this but when I run the sample I often see messages that are delayed by more than 20s - does the timeout apply to websockets as well?

What's the bebusy() function for on the client side? Does the delay reproduce without that? How many clients are you using for this test? How can we replicate this test on our hardware?

**Comment by davidfowl** [comment]
I'm more convinced now that you should introduce your own buffer to accomplish what you want. Treat SignalR like the transport layer instead of the messaging layer. You're flooding the transport (TCP) with lots of data and the side effect is back pressure [which you're are ignoring in this case](https://github.com/jfheins/SignalRRepro/blob/b477cd90bdd93e1efaa6633252347c3bfcc8213c/MessageSender.cs#L37). SignalR does not drop messages and the timeout at the transport layer (the one described in my initial reply) is a mitigation to stop hostile actors from purposely slowing down the transport layer (by not reading data).

Controlling memory usage means in the face of back pressure because you're hit some threshold:
1. In that sense you need to keep a buffer of relevant messages for SignalR.
2. Dropping messages 

SignalR doesn't drop messages (unless you cancelled the call), so if this is the behavior you want, I'd suggest some changes to your code to keep a handle to the most recent message since they aren't important. Also make the change to respect back pressure when sending messages to clients.

Once you do that, you can play around with buffer sizes to see what gives the most satisfactory memory usage. Specifically [this one](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connections.httpconnectiondispatcheroptions.applicationmaxbuffersize?view=aspnetcore-6.0#microsoft-aspnetcore-http-connections-httpconnectiondispatcheroptions-applicationmaxbuffersize)

**Comment by ArminShoeibi** [comment]
Hi there @jfheins, What is your application GC mode?
If you are running dotnet in containers you should change your GC mode.
<img width="531" alt="image" src="https://user-images.githubusercontent.com/68565441/183758355-7f9c37e6-19ac-4229-898a-8aae5a526dd2.png">


> Over 99% of the messages are state messages, i.e. as soon as a newer state is known, all older states of the same widget are irrelevent. On the backend (.net 6 Application with Kestrel in a K8s cluster, connected via websockets) the memory consumption grows linearly when such a slow client id connected:

**Comment by davidfowl** [comment]
@jfheins I sent a PR to show how you would await backpressure here https://github.com/jfheins/SignalRRepro/pull/1.

Also to @ArminShoeibi's point, you should overlay garbage collections with the memory growth.

**Comment by ArminShoeibi** [comment]
> @jfheins I sent a PR to show how you would await backpressure here [jfheins/SignalRRepro#1](https://github.com/jfheins/SignalRRepro/pull/1).
> 
> Also to @ArminShoeibi's point, you should overlay garbage collections with the memory growth.

Thanks David, I've rewritten this code. like this but I have a memory leak.
<img width="520" alt="image" src="https://user-images.githubusercontent.com/68565441/183763021-ede25722-1cda-4cf0-ba0d-a35fe045fdd7.png">

**Comment by davidfowl** [comment]
What's the leak?

**Comment by davidfowl** [comment]
The cancellation token shouldn't be reused and should instead be disposed (or you can reset it).

**Comment by davidfowl** [comment]
After running this sample a few times I'm convinced that the memory growth is just because there's been no GCs happening. I've run it under an allocation profiler for an hour (mostly because I forgot it was running 😄) and there were no GCs because it allocated so little. My machine has LOTS of cores though, so I'd like to run the runtime under whatever configuration container limits you were using.

@jfheins can you share how your container limits were configured?

PS: The back pressure situation should still be fixed.

**Comment by ArminShoeibi** [comment]
> The cancellation token shouldn't be reused and should instead be disposed (or you can reset it).

I got it.

**Comment by ivanthelad** [comment]
> ecause I forgot it was running 😄) and there were no GCs because it allocated so little. My machine has LOTS of cores though, so I'd like to run the runtime under whatever configuration container limits you were using.
> 
> @jfheins can you s

@davidfowl  did you configure any gc limit during your test

**Comment by jfheins** [comment]
I have experimented with the cancellation token a bit more and saw that the message delay can reach about 5x the token time. So with a token of 1s I saw message delay of 5s.

**Comment by davidfowl** [comment]
OK I've been playing with this for a bit and think I've come to some conclusions:
- Buffering is done for performance reasons (as you can imagine), so when you call Clients.All.SendAsync, it's writing into a local buffer in each client and returning as fast as it can. That data is then read by the transport and written there (which has another buffer etc etc all the way to the TLS layer and eventually the OS tcp buffer). There needs to be back pressure in these layers for the cancellation token to do anything useful. So when SendAsync returns, it's put data into each of the buffers of the clients, you can make this very fast by having a short lived cancellation token but that will basically drop slow clients (completely disconnect them), it will not help with making them catch up to new messages.
- The cancellation token will make sure that slow clients don't affect faster clients or the entire broadcast, but there's no way to make slow clients drop older messages and catch up. The token will end up dropping clients that hit the threshold, so they'll drop and reconnect.
- The memory in the server side should be fine with or without slow clients. I'm happy to hear that running an induced GC (via GC.Collect) reclaimed the growing memory, it should do that.

Here's my latest change that use a global buffer
https://github.com/jfheins/SignalRRepro/pull/2/files#diff-914617f2e5e469273ca2927ab60578f11dacd836d2a02f5e2189b6e8fbe01c61

The only way to make slow clients catch up is to keep track of all clients yourself, keep a buffer per clients, and send messages to that client using that buffer. I play to write that sample next.

**Comment by davidfowl** [comment]
PS:I was wrong about the setting that controls back pressure on the transport side 🤦🏾‍♂️. It's actually https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.http.connections.httpconnectiondispatcheroptions.transportmaxbuffersize?view=aspnetcore-6.0#microsoft-aspnetcore-http-connections-httpconnectiondispatcheroptions-transportmaxbuffersize. Though I'm having a hard time getting it to work well in my current spike.

**Comment by jfheins** [comment]
Thanks for the engagement, it really makes me feel that our issue is heard and we're enabled to make more cool software 😊

First off maybe the main reason I chose that title: When the memory accumulates, we are getting OOM exceptions. Having read more about the server GC, I understand that memory accumulation itself is intended behavior, with the GC kicking in as soon as the managed memory exceeds 75% of the pod limit. Therefore my best guess is that a significant chunk of that memory is unmanaged so that the GC is not triggered, yet the overall memory is insufficient. We can tune the behavior using env vars to set the limit differently or workstation GC 👍

> Here's my latest change that use a global buffer

 Thank you a lot, I will have a deeper look into that. The blocker for prod would be that all clients loose messages when one is slow.

> The only way to make slow clients catch up is to keep track of all clients yourself, keep a buffer per clients, and send messages to that client using that buffer. I plan to write that sample next.

That might well be what we have to do then. It would be awesome to have a sample there. My idea so far would be then to minimize message staleness by setting the buffer to 10KB (large enough for the largest message + margin) and essentially discard a few messages as soon as the send task is not "IsCompletedSuccessfully" immediately. (Here a "SendNowOrNever" method would still be nice 😇)

**Comment by davidfowl** [comment]
I’ve been working on this issue for 3 nights now (it’s completely nerd sniped me). I haven’t run with the limits yet but I’ve been focusing on the delay client side not the server side. I’ll share what I have and I’ve been debugging. I’ve been playing with configuring the buffers at all layers to minimize the delay and it’s still pretty unpredictable.

**Comment by davidfowl** [comment]
One last question, can you reduce the message rate? Is that also an option here?

**Comment by jfheins** [comment]
> One last question, can you reduce the message rate? Is that also an option here?

I am currently working on some batching, i.e. collecting messages over 50nms and sending an array. (The idea is that fewer, larger messages have less client overhead than many smaller ones)
The next idea is to send information about the viewport to the server and then filter the messages server side. Then users would have less data to process if they are zoomed in. For zoomed out users, we could reduce data fidelity. We did not do that (yet) because we feared a lot of complex (per-client) logic in the backend. (We would have to generate "this widget left your viewport" events)
And with this option compared to polling via SignalR, I am still leaning towards polling. (With polling there should be no queue so the message delay should be minimal and the rate is set by the client, so it ticks many boxes already)

**Comment by davidfowl** [comment]
As promised https://github.com/jfheins/SignalRRepro/compare/main...davidfowl:SignalRRepro:davidfowl/channel-per-conn?expand=1

**Comment by jfheins** [comment]
> As promised

I gave it a test run 👍
There is still plenty of delay building up in the client, but the memory server side looks very good (=constant) 💯 

https://user-images.githubusercontent.com/5293502/184552957-50954eb5-fcfa-4918-91f4-1fa84198a1c7.mp4

Btw. when I open 2 tabs, I have the impression that only one of them get messages at any given instant. (I'm running it now on a desktop machine, not on my work laptop).

**Comment by davidfowl** [comment]
I see this behavior too and can’t explain it as yet. I want to verify that behavior on with the clients and servers on different compute.

**Comment by davidfowl** [comment]
I think the only thing that could be done in this situation would be do:
- Reduce server side the message rate, the client cannot keep up so the server needs to send less often to the clients.
- Reduce the delay on the client by having the client send an act back to the server with the last ID it saw before sending the next message.

**Comment by davidfowl** [comment]
OK it's solved! I had to update to .NET 7 to make this easier since we just added a new feature that makes this work (client results). That basically waits for the client to ack that it processed a message before moving to the next one. 

See https://github.com/jfheins/SignalRRepro/commit/1d10044da9f601da53a20618b25f3d56f090a73b

https://user-images.githubusercontent.com/95136/184661346-28e95ea5-260d-4447-91a9-39077e672326.mp4
